src/main/java/software/amazon/cloudwatchlogs/emf/sinks/AgentSink.java

```diff
@@ -17,28 +17,61 @@
 package software.amazon.cloudwatchlogs.emf.sinks;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import lombok.AllArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import software.amazon.cloudwatchlogs.emf.Constants;
+import software.amazon.cloudwatchlogs.emf.exception.EMFClientException;
 import software.amazon.cloudwatchlogs.emf.model.MetricsContext;
+import software.amazon.cloudwatchlogs.emf.sinks.retry.RetryStrategy;
 import software.amazon.cloudwatchlogs.emf.util.StringUtils;
 
-/** An sink connecting to CloudWatch Agent. */
+/** An sink connecting to an agent over a socket. */
 @Slf4j
 public class AgentSink implements ISink {
     private final String logGroupName;
     private final String logStreamName;
     private final SocketClient client;
+    private final ExecutorService executor;
+    private final Supplier<RetryStrategy> retryStrategyFactory;
+    private final LinkedBlockingQueue<Runnable> queue;
 
     public AgentSink(
             String logGroupName,
             String logStreamName,
             Endpoint endpoint,
-            SocketClientFactory clientFactory) {
+            SocketClientFactory clientFactory,
+            int asyncQueueDepth,
+            Supplier<RetryStrategy> retryStrategy) {
         this.logGroupName = logGroupName;
         this.logStreamName = logStreamName;
         client = clientFactory.getClient(endpoint);
+        queue = new LinkedBlockingQueue<>(asyncQueueDepth);
+        executor = createSingleThreadedExecutor();
+        this.retryStrategyFactory = retryStrategy;
+    }
+
+    private ExecutorService createSingleThreadedExecutor() {
+        return new ThreadPoolExecutor(
+                1,
+                1,
+                0L,
+                TimeUnit.MILLISECONDS,
+                queue,
+                new ThreadPoolExecutor.DiscardOldestPolicy());
```

Comment:
TODO: add test on ring buffer behavior (should we be using and actual ring buffer instead of this rejection policy?)